### PR TITLE
- AI: Reduce priority of mana sources that make you sacrifice other t…

### LINKED ISSUE #3156: Feature Request: Remove Ashnod's Altar/Phyrexian Altar from autotapper
**Is your feature request related to a problem? Please describe.**
Most of the time, since I'm aware of the issue, I try to manually tap for mana when I have either Ashnod's Altar or Phyrexian Altar out, but sometimes (often) you just forget... The sac'ing of creatures when other options are *always* better is really bad, like: "I have to exit the game bad, 'cause Forge just ended my board." 

The most egregious just happened a few days ago, which is what prompted me to try to ask for this in the first place. I had a very strong board and six treasures ready to be tapped, plus Phyrexian Altar. I, thinking that surely Forge would be smart enough to tap and sac the obvious SIX treasures for mana when I cast a 6CMC card, didn't even think about being warry of the autotapper... The game sac'd all six of my creatures to Phyrexian Altar instead of using the six treasures. I had to quit.

**Describe the solution you'd like**
Just completely remove Ashnod's Altar and Phyrexian Altar from any kind of autotapping solution. Just don't allow it. If you don't allow it, then Forge can't kill your creatures due to a user's innocent "error" when other much better options are present. 

I know I can't be the only person who has had this issue, so why not just resolve this once and for all? I don't know how hard it would be, maybe very hard, but leaving an obvious problem in and just saying "it's your fault for not tapping manually" is kinda not cool.

Anywho, I really, really do love Forge and I use it a lot. I understand that this is all volunteer and no one gets paid for this project of love, but it *is* a project of love, so maybe show a little bit more love and fix this frustration/rage inducing interaction?

### PR DESCRIPTION
**Pull request by tehdiplomat**
…hings

Closes #3156

### CODE REVIEW

**Comment by tool4ever** [comment]
Will review tonight :)

**Comment by Hanmac** [comment]
i don't know if that already handles it,
but can the Auto Payment (and AI) prioritize Treasure token over Treasure creatures like Gold Hound?

File: forge-ai/src/main/java/forge/ai/ComputerUtilMana.java [hunk 1]
```diff
@@ -1521,7 +1523,17 @@ public boolean apply(final Card c) {
...
 
                 if (cost != null) {
-                    needsLimitedResources |= !cost.isReusuableResource();
+                    if (!cost.isReusuableResource()) {
+                        for(CostPart part : cost.getCostParts()) {
+                            if (part instanceof CostSacrifice) {
+                                if (!part.getType().equals("CARDNAME")) {
+                                    unpreferredCost = true;
+                                } else {
+                                    needsLimitedResources = true;
```
Comment (tool4ever) [inline forge-ai/src/main/java/forge/ai/ComputerUtilMana.java:1532]: it seems the logic got "degraded" a bit here:
now `needsLimitedResources` can only be true if there's a Sac Cost which seems wrong

File: forge-ai/src/main/java/forge/ai/ComputerUtilMana.java [hunk 2]
```diff
@@ -1521,7 +1523,17 @@ public boolean apply(final Card c) {
                 final Cost cost = m.getPayCosts();
 
                 if (cost != null) {
-                    needsLimitedResources |= !cost.isReusuableResource();
+                    if (!cost.isReusuableResource()) {
```
Comment (tool4ever) [inline forge-ai/src/main/java/forge/ai/ComputerUtilMana.java:1526]: `ComputerUtilCard.isSacrificeSelfCost`

File: forge-game/src/main/java/forge/game/spellability/SpellAbility.java [hunk 3]
```diff
@@ -2284,6 +2285,11 @@ public int calculateScoreForManaAbility() {
             if (costPart instanceof CostTap && !Untap.canUntap(getHostCard())) {
                 score += 10;
             }
+            if (costPart instanceof CostSacrifice && !costPart.getType().equals("CARDNAME")) {
```
Comment (tool4ever) [inline forge-game/src/main/java/forge/game/spellability/SpellAbility.java:2288]: `costPart.payCostFromSource()`

File: forge-ai/src/main/java/forge/ai/ComputerUtilMana.java (hunk 2 above)
Comment (tehdiplomat) [reply forge-ai/src/main/java/forge/ai/ComputerUtilMana.java:1526]: That function doesn't really help me. I need to check if its a sacrifice Cost AND not a self sacrifice cost.

File: forge-ai/src/main/java/forge/ai/ComputerUtilMana.java (hunk 1 above)
Comment (tehdiplomat) [reply forge-ai/src/main/java/forge/ai/ComputerUtilMana.java:1532]: Ah yea, Lemme refactor a bit to get what I need.

**Review by tool4ever — APPROVED** [review]

**Pull request merged**